fuse_models/include/fuse_models/common/throttled_callback.h

```diff
@@ -0,0 +1,159 @@
+
+#include <ros/subscriber.h>
+
+#include <functional>
+
+
+namespace fuse_models
+{
+
+namespace common
+{
+
+/**
+ * @brief A throttled callback that encapsulates the logic to throttle a message callback so it is only called after a
+ * given period in seconds (or more). The dropped messages can optionally be received in a dropped callback, that could
+ * be used to count the number of messages dropped.
+ *
+ * @tparam M The message the callback receives
+ */
+template <class M>
+class ThrottledCallback
+{
+public:
+  using MessageConstPtr = typename M::ConstPtr;
+  using Callback = std::function<void(const MessageConstPtr&)>;
+
+  /**
+   * @brief Constructor
+   *
+   * @param[in] keep_callback   The callback to call when the message is kept, i.e. not dropped. Defaults to nullptr
+   * @param[in] drop_callback   The callback to call when the message is dropped because of the throttling. Defaults to
+   *                            nullptr
+   * @param[in] throttle_period The throttling period duration in seconds. Defaults to 0.0, i.e. no throttling
+   */
+  ThrottledCallback(Callback&& keep_callback = nullptr,  // NOLINT(whitespace/operators)
+                    Callback&& drop_callback = nullptr,  // NOLINT(whitespace/operators)
+                    const ros::Duration& throttle_period = ros::Duration(0.0))
+    : keep_callback_(keep_callback), drop_callback_(drop_callback), throttle_period_(throttle_period)
+  {
+  }
+
+  /**
+   * @brief Throttle period getter
+   *
+   * @return The current throttle period duration in seconds being used
+   */
+  const ros::Duration& getThrottlePeriod() const
+  {
+    return throttle_period_;
+  }
+
+  /**
+   * @brief Throttle period setter
+   *
+   * @param[in] throttle_period The new throttle period duration in seconds to use
+   */
+  void setThrottlePeriod(const ros::Duration& throttle_period)
+  {
+    throttle_period_ = throttle_period;
+  }
+
+  /**
+   * @brief Keep callback setter
+   *
+   * @param[in] keep_callback The new keep callback to use
+   */
+  void setKeepCallback(const Callback& keep_callback)
+  {
+    keep_callback_ = keep_callback;
+  }
+
+  /**
+   * @brief Drop callback setter
+   *
+   * @param[in] drop_callback The new drop callback to use
+   */
+  void setDropCallback(const Callback& drop_callback)
+  {
+    drop_callback_ = drop_callback;
+  }
+
+  /**
+   * @brief Last called time
+   *
+   * @return The last time the keep callback was called
+   */
+  const ros::Time& getLastCalledTime() const
+  {
+    return last_called_time_;
+  }
+
+  /**
+   * @brief Message callback that throttles the calls to the keep callback provided. When the message is dropped because
+   * of throttling, the drop callback is called instead.
+   *
+   * @param[in] message The input message
+   */
+  void callback(const MessageConstPtr& message)
+  {
+    const auto now = ros::Time::now();
+    if (!last_called_time_.isValid() || now - last_called_time_ > throttle_period_)
```

Comment:
In case it's not clear, the `!last_called_timer_.isValid()` check is done in order to always published the first message received, even if `now` is very close to `0`, which could happen in simulation or in bag recorded from simulation.

I also wonder if I should provide a ROS param `use_wall_time` to use the `ros::WallTime::now()` instead of the `ros::Time::now()`, similar to `topic_tools throttle`:
https://github.com/ros/ros_comm/blob/d18e0dd7a7ffe5b76d5d31a9e1ba7f150e486f73/tools/topic_tools/src/throttle.cpp#L124-L138

Reply:
You can add a walltime option if you want; it definitely doesn't hurt anything. I'm just not sure what the use-case is. If I'm running a sim at 0.5x or 5x realtime, I still want the sensor data to behave the same.

Reply:
I just added a `use_wall_time_` flag and `throttle_use_wall_time` (not sure about the name :thinking:) ROS param. Both default to `false`.

I think when we play things back from a bag file, it could be useful to use the `ros::WallTime`, so everything gets processed if we either play at a low rate, or we pause playing. Indeed, when we pause playing something special happens: if the callback queue has pending callbacks, they keep getting processed, even if the `ros::Time` is not growing. Indeed, this is a corner case not handled well with the current implementation, because the `now - last_called_time_` difference would be `0.0`, and those callbacks get dropped, even if `throttle_period_ == 0.0`. :disappointed: 

I could either handle `throttle_period_ == 0.0` differently, or allow `throttle_period_ < 0.0`, and use `-1.0` as a default. But actually, `0.0` should be handled properly, so it looks like the right answer here would be to handle that value differently, as a particular special case, where all callbacks are kept regardless of `now` and `last_called_time_`.

Reply:
I've updated this PR with this. I've also rebased it on top of #169 , which now must be review and merged before continue reviewing this one.